无法从 dict 反序列化为 Message

如以下的代码：
```py
def test_message_deserialize():
    """测试消息反序列化"""
    from nonebot.adapters.qq import Message, MessageSegment
    from nonebot.compat import type_validate_python

    msg = type_validate_python(Message, [{"type": "text", "data": {"text": "test"}}])
    assert msg == Message(MessageSegment.text("test"))
```

会出现如下报错：
```
_________________________________________ test_message_deserialize _________________________________________

    def test_message_deserialize():
        """测试消息反序列化"""
        from nonebot.adapters.qq import Message, MessageSegment
        from nonebot.compat import type_validate_python

>       msg = type_validate_python(Message, [{"type": "text", "data": {"text": "test"}}])

tests/test_qq.py:338:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.10/site-packages/nonebot/compat.py:225: in type_validate_python
    return TypeAdapter(type_).validate_python(data)
.venv/lib/python3.10/site-packages/pydantic/type_adapter.py:135: in wrapped
    return func(self, *args, **kwargs)
.venv/lib/python3.10/site-packages/pydantic/type_adapter.py:366: in validate_python
    return self.validator.validate_python(object, strict=strict, from_attributes=from_attributes, context=context)
.venv/lib/python3.10/site-packages/nonebot/internal/adapter/message.py:166: in _validate
    value = [type_validate_python(cls.get_segment_class(), v) for v in value]
.venv/lib/python3.10/site-packages/nonebot/internal/adapter/message.py:166: in <listcomp>
    value = [type_validate_python(cls.get_segment_class(), v) for v in value]
.venv/lib/python3.10/site-packages/nonebot/compat.py:225: in type_validate_python
    return TypeAdapter(type_).validate_python(data)
.venv/lib/python3.10/site-packages/pydantic/type_adapter.py:135: in wrapped
    return func(self, *args, **kwargs)
.venv/lib/python3.10/site-packages/pydantic/type_adapter.py:366: in validate_python
    return self.validator.validate_python(object, strict=strict, from_attributes=from_attributes, context=context)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'nonebot.adapters.qq.message.MessageSegment'>
value = {'data': {'text': 'test'}, 'type': 'text'}

    @classmethod
    def _validate(cls, value) -> Self:
        if isinstance(value, cls):
            return value
        if isinstance(value, MessageSegment):
            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
        if not isinstance(value, dict):
            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
        if "type" not in value:
            raise ValueError(
                f"Expected dict with 'type' for MessageSegment, got {value}"
            )
>       return cls(type=value["type"], data=value.get("data", {}))
E       TypeError: Can't instantiate abstract class MessageSegment with abstract method __str__

.venv/lib/python3.10/site-packages/nonebot/internal/adapter/message.py:76: TypeError
```
预期结果：
希望如下关系成立：
```py
type_validate_python(Message, [{"type": "text", "data": {"text": "test"}}]) == Message(MessageSegment.text("test"))
```

--- a/nonebot/adapters/qq/message.py
+++ b/nonebot/adapters/qq/message.py
@@ -14,7 +14,6 @@ from typing import (
     overload,
 )
 
-from pydantic import BaseModel
 from nonebot.compat import type_validate_python
 
 from nonebot.adapters import Message as BaseMessage
@@ -194,11 +193,8 @@ class MessageSegment(BaseMessageSegment["Message"]):
         _type = value["type"]
         if _type not in SEGMENT_TYPE_MAP:
             raise ValueError(f"Invalid MessageSegment type: {_type}")
-        data = value.get("data", {})
-        for key, data_type in SEGMENT_DATA_MAP.items():
-            if key in data and not isinstance(data[key], data_type):
-                data[key] = type_validate_python(data_type, data[key])
-        return SEGMENT_TYPE_MAP[_type](type=_type, data=data)
+        sub_cls = SEGMENT_TYPE_MAP[_type]
+        return sub_cls._validate(sub_cls(type=_type, data=value.get("data", {})))
 
 
 class _TextData(TypedDict):
@@ -312,6 +308,15 @@ class Embed(MessageSegment):
     def __str__(self) -> str:
         return f"<embed:{self.data['embed']!r}>"
 
+    @classmethod
+    @override
+    def _validate(cls, value) -> Self:
+        if not isinstance(value, cls):
+            value = super()._validate(value)
+        if not isinstance(embed := value.data["embed"], MessageEmbed):
+            value.data["embed"] = type_validate_python(MessageEmbed, embed)
+        return value
+
 
 class _ArkData(TypedDict):
     ark: MessageArk
@@ -325,6 +330,15 @@ class Ark(MessageSegment):
     @override
     def __str__(self) -> str:
         return f"<ark:{self.data['ark']!r}>"
+
+    @classmethod
+    @override
+    def _validate(cls, value) -> Self:
+        if not isinstance(value, cls):
+            value = super()._validate(value)
+        if not isinstance(ark := value.data["ark"], MessageArk):
+            value.data["ark"] = type_validate_python(MessageArk, ark)
+        return value
 
 
 class _ReferenceData(TypedDict):
@@ -340,6 +354,15 @@ class Reference(MessageSegment):
     def __str__(self) -> str:
         return f"<reference:{self.data['reference'].message_id}>"
 
+    @classmethod
+    @override
+    def _validate(cls, value) -> Self:
+        if not isinstance(value, cls):
+            value = super()._validate(value)
+        if not isinstance(reference := value.data["reference"], MessageReference):
+            value.data["reference"] = type_validate_python(MessageReference, reference)
+        return value
+
 
 class _MarkdownData(TypedDict):
     markdown: MessageMarkdown
@@ -354,6 +377,15 @@ class Markdown(MessageSegment):
     def __str__(self) -> str:
         return f"<markdown:{self.data['markdown']!r}>"
 
+    @classmethod
+    @override
+    def _validate(cls, value) -> Self:
+        if not isinstance(value, cls):
+            value = super()._validate(value)
+        if not isinstance(markdown := value.data["markdown"], MessageMarkdown):
+            value.data["markdown"] = type_validate_python(MessageMarkdown, markdown)
+        return value
+
 
 class _KeyboardData(TypedDict):
     keyboard: MessageKeyboard
@@ -367,6 +399,15 @@ class Keyboard(MessageSegment):
     @override
     def __str__(self) -> str:
         return f"<keyboard:{self.data['keyboard']!r}>"
+
+    @classmethod
+    @override
+    def _validate(cls, value) -> Self:
+        if not isinstance(value, cls):
+            value = super()._validate(value)
+        if not isinstance(keyboard := value.data["keyboard"], MessageKeyboard):
+            value.data["keyboard"] = type_validate_python(MessageKeyboard, keyboard)
+        return value
 
 
 SEGMENT_TYPE_MAP: Dict[str, Type[MessageSegment]] = {
@@ -388,14 +429,6 @@ SEGMENT_TYPE_MAP: Dict[str, Type[MessageSegment]] = {
     "markdown": Markdown,
     "keyboard": Keyboard,
     "reference": Reference,
-}
-
-SEGMENT_DATA_MAP: Dict[str, Type[BaseModel]] = {
-    "embed": MessageEmbed,
-    "ark": MessageArk,
-    "reference": MessageReference,
-    "markdown": MessageMarkdown,
-    "keyboard": MessageKeyboard,
 }
 
 

--- a/nonebot/adapters/qq/message.py
+++ b/nonebot/adapters/qq/message.py
@@ -193,8 +193,12 @@ class MessageSegment(BaseMessageSegment["Message"]):
         _type = value["type"]
         if _type not in SEGMENT_TYPE_MAP:
             raise ValueError(f"Invalid MessageSegment type: {_type}")
-        sub_cls = SEGMENT_TYPE_MAP[_type]
-        return sub_cls._validate(sub_cls(type=_type, data=value.get("data", {})))
+        segment_type = SEGMENT_TYPE_MAP[_type]
+        if cls is MessageSegment:
+            return type_validate_python(segment_type, value)
+        if cls is segment_type:
+            return segment_type(type=_type, data=value.get("data", {}))
+        raise ValueError(f"Type {cls} can not be converted to {segment_type}")
 
 
 class _TextData(TypedDict):
@@ -311,11 +315,21 @@ class Embed(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if not isinstance(value, cls):
-            value = super()._validate(value)
-        if not isinstance(embed := value.data["embed"], MessageEmbed):
-            value.data["embed"] = type_validate_python(MessageEmbed, embed)
-        return value
+        if isinstance(value, cls):
+            return value
+        if isinstance(value, MessageSegment):
+            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
+        if not isinstance(value, dict):
+            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
+        if "data" not in value:
+            raise ValueError(f"Expected dict with 'data' for Embed, got {value}")
+        if "embed" not in value["data"]:
+            raise ValueError(
+                f"Expected dict with 'embed' in 'data' for Embed, got {value}"
+            )
+        if not isinstance(embed := value["data"]["embed"], MessageEmbed):
+            value["data"]["embed"] = type_validate_python(MessageEmbed, embed)
+        return cls(type="embed", data=value["data"])
 
 
 class _ArkData(TypedDict):
@@ -334,11 +348,19 @@ class Ark(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if not isinstance(value, cls):
-            value = super()._validate(value)
-        if not isinstance(ark := value.data["ark"], MessageArk):
-            value.data["ark"] = type_validate_python(MessageArk, ark)
-        return value
+        if isinstance(value, cls):
+            return value
+        if isinstance(value, MessageSegment):
+            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
+        if not isinstance(value, dict):
+            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
+        if "data" not in value:
+            raise ValueError(f"Expected dict with 'data' for Ark, got {value}")
+        if "ark" not in value["data"]:
+            raise ValueError(f"Expected dict with 'ark' in 'data' for Ark, got {value}")
+        if not isinstance(ark := value["data"]["ark"], MessageArk):
+            value["data"]["ark"] = type_validate_python(MessageArk, ark)
+        return cls(type="ark", data=value["data"])
 
 
 class _ReferenceData(TypedDict):
@@ -357,11 +379,23 @@ class Reference(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if not isinstance(value, cls):
-            value = super()._validate(value)
-        if not isinstance(reference := value.data["reference"], MessageReference):
-            value.data["reference"] = type_validate_python(MessageReference, reference)
-        return value
+        if isinstance(value, cls):
+            return value
+        if isinstance(value, MessageSegment):
+            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
+        if not isinstance(value, dict):
+            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
+        if "data" not in value:
+            raise ValueError(f"Expected dict with 'data' for Reference, got {value}")
+        if "reference" not in value["data"]:
+            raise ValueError(
+                f"Expected dict with 'reference' in 'data' for Reference, got {value}"
+            )
+        if not isinstance(reference := value["data"]["reference"], MessageReference):
+            value["data"]["reference"] = type_validate_python(
+                MessageReference, reference
+            )
+        return cls(type="reference", data=value["data"])
 
 
 class _MarkdownData(TypedDict):
@@ -380,11 +414,21 @@ class Markdown(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if not isinstance(value, cls):
-            value = super()._validate(value)
-        if not isinstance(markdown := value.data["markdown"], MessageMarkdown):
-            value.data["markdown"] = type_validate_python(MessageMarkdown, markdown)
-        return value
+        if isinstance(value, cls):
+            return value
+        if isinstance(value, MessageSegment):
+            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
+        if not isinstance(value, dict):
+            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
+        if "data" not in value:
+            raise ValueError(f"Expected dict with 'data' for Markdown, got {value}")
+        if "markdown" not in value["data"]:
+            raise ValueError(
+                f"Expected dict with 'markdown' in 'data' for Markdown, got {value}"
+            )
+        if not isinstance(markdown := value["data"]["markdown"], MessageMarkdown):
+            value["data"]["markdown"] = type_validate_python(MessageMarkdown, markdown)
+        return cls(type="markdown", data=value["data"])
 
 
 class _KeyboardData(TypedDict):
@@ -403,11 +447,21 @@ class Keyboard(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if not isinstance(value, cls):
-            value = super()._validate(value)
-        if not isinstance(keyboard := value.data["keyboard"], MessageKeyboard):
-            value.data["keyboard"] = type_validate_python(MessageKeyboard, keyboard)
-        return value
+        if isinstance(value, cls):
+            return value
+        if isinstance(value, MessageSegment):
+            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
+        if not isinstance(value, dict):
+            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
+        if "data" not in value:
+            raise ValueError(f"Expected dict with 'data' for Keyboard, got {value}")
+        if "keyboard" not in value["data"]:
+            raise ValueError(
+                f"Expected dict with 'keyboard' in 'data' for Keyboard, got {value}"
+            )
+        if not isinstance(keyboard := value["data"]["keyboard"], MessageKeyboard):
+            value["data"]["keyboard"] = type_validate_python(MessageKeyboard, keyboard)
+        return cls(type="keyboard", data=value["data"])
 
 
 SEGMENT_TYPE_MAP: Dict[str, Type[MessageSegment]] = {

--- a/nonebot/adapters/qq/message.py
+++ b/nonebot/adapters/qq/message.py
@@ -1,10 +1,21 @@
 import re
 from io import BytesIO
 from pathlib import Path
+from pydantic import BaseModel
 from dataclasses import dataclass
 from typing_extensions import Self, override
-from typing import TYPE_CHECKING, Type, Union, Iterable, Optional, TypedDict, overload
+from typing import (
+    TYPE_CHECKING,
+    Type,
+    Union,
+    Iterable,
+    Optional,
+    TypedDict,
+    overload,
+    Dict,
+)
 
+from nonebot.compat import type_validate_python
 from nonebot.adapters import Message as BaseMessage
 from nonebot.adapters import MessageSegment as BaseMessageSegment
 
@@ -179,11 +190,14 @@ class MessageSegment(BaseMessageSegment["Message"]):
             raise ValueError(
                 f"Expected dict with 'type' for MessageSegment, got {value}"
             )
-        if value["type"] not in SEGMENT_TYPE_MAP:
-            raise ValueError(f"Invalid MessageSegment type: {value['type']}")
-        return SEGMENT_TYPE_MAP[value["type"]](
-            type=value["type"], data=value.get("data", {})
-        )
+        _type = value["type"]
+        if _type not in SEGMENT_TYPE_MAP:
+            raise ValueError(f"Invalid MessageSegment type: {_type}")
+        data = value.get("data", {})
+        for key, data_type in SEGMENT_DATA_MAP.items():
+            if key in data and not isinstance(data[key], data_type):
+                data[key] = type_validate_python(data_type, data[key])
+        return SEGMENT_TYPE_MAP[_type](type=_type, data=data)
 
 
 class _TextData(TypedDict):
@@ -354,7 +368,7 @@ class Keyboard(MessageSegment):
         return f"<keyboard:{self.data['keyboard']!r}>"
 
 
-SEGMENT_TYPE_MAP = {
+SEGMENT_TYPE_MAP: Dict[str, Type[MessageSegment]] = {
     "text": Text,
     "emoji": Emoji,
     "mention_user": MentionUser,
@@ -373,6 +387,14 @@ SEGMENT_TYPE_MAP = {
     "markdown": Markdown,
     "keyboard": Keyboard,
     "reference": Reference,
+}
+
+SEGMENT_DATA_MAP: Dict[str, Type[BaseModel]] = {
+    "embed": MessageEmbed,
+    "ark": MessageArk,
+    "reference": MessageReference,
+    "markdown": MessageMarkdown,
+    "keyboard": MessageKeyboard,
 }
 
 

--- a/nonebot/adapters/qq/message.py
+++ b/nonebot/adapters/qq/message.py
@@ -1,21 +1,22 @@
 import re
 from io import BytesIO
 from pathlib import Path
-from pydantic import BaseModel
 from dataclasses import dataclass
 from typing_extensions import Self, override
 from typing import (
     TYPE_CHECKING,
+    Dict,
     Type,
     Union,
     Iterable,
     Optional,
     TypedDict,
     overload,
-    Dict,
 )
 
+from pydantic import BaseModel
 from nonebot.compat import type_validate_python
+
 from nonebot.adapters import Message as BaseMessage
 from nonebot.adapters import MessageSegment as BaseMessageSegment
 

--- a/nonebot/adapters/qq/message.py
+++ b/nonebot/adapters/qq/message.py
@@ -166,6 +166,25 @@ class MessageSegment(BaseMessageSegment["Message"]):
     def is_text(self) -> bool:
         return self.type == "text"
 
+    @classmethod
+    @override
+    def _validate(cls, value) -> Self:
+        if isinstance(value, cls):
+            return value
+        if isinstance(value, MessageSegment):
+            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
+        if not isinstance(value, dict):
+            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
+        if "type" not in value:
+            raise ValueError(
+                f"Expected dict with 'type' for MessageSegment, got {value}"
+            )
+        if value["type"] not in SEGMENT_TYPE_MAP:
+            raise ValueError(f"Invalid MessageSegment type: {value['type']}")
+        return SEGMENT_TYPE_MAP[value["type"]](
+            type=value["type"], data=value.get("data", {})
+        )
+
 
 class _TextData(TypedDict):
     text: str
@@ -333,6 +352,28 @@ class Keyboard(MessageSegment):
     @override
     def __str__(self) -> str:
         return f"<keyboard:{self.data['keyboard']!r}>"
+
+
+SEGMENT_TYPE_MAP = {
+    "text": Text,
+    "emoji": Emoji,
+    "mention_user": MentionUser,
+    "mention_channel": MentionChannel,
+    "mention_everyone": MentionEveryone,
+    "image": Attachment,
+    "file_image": LocalAttachment,
+    "audio": Attachment,
+    "file_audio": LocalAttachment,
+    "video": Attachment,
+    "file_video": LocalAttachment,
+    "file": Attachment,
+    "file_file": LocalAttachment,
+    "ark": Ark,
+    "embed": Embed,
+    "markdown": Markdown,
+    "keyboard": Keyboard,
+    "reference": Reference,
+}
 
 
 class Message(BaseMessage[MessageSegment]):

--- a/nonebot/adapters/qq/message.py
+++ b/nonebot/adapters/qq/message.py
@@ -194,11 +194,14 @@ class MessageSegment(BaseMessageSegment["Message"]):
         if _type not in SEGMENT_TYPE_MAP:
             raise ValueError(f"Invalid MessageSegment type: {_type}")
         segment_type = SEGMENT_TYPE_MAP[_type]
+
+        # casting value to subclass of MessageSegment
         if cls is MessageSegment:
             return type_validate_python(segment_type, value)
+        # init segment instance directly if type matched
         if cls is segment_type:
             return segment_type(type=_type, data=value.get("data", {}))
-        raise ValueError(f"Type {cls} can not be converted to {segment_type}")
+        raise ValueError(f"Segment type {_type!r} can not be converted to {cls}")
 
 
 class _TextData(TypedDict):
@@ -315,21 +318,14 @@ class Embed(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if isinstance(value, cls):
-            return value
-        if isinstance(value, MessageSegment):
-            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
-        if not isinstance(value, dict):
-            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
-        if "data" not in value:
-            raise ValueError(f"Expected dict with 'data' for Embed, got {value}")
-        if "embed" not in value["data"]:
+        instance = super()._validate(value)
+        if "embed" not in instance.data:
             raise ValueError(
                 f"Expected dict with 'embed' in 'data' for Embed, got {value}"
             )
-        if not isinstance(embed := value["data"]["embed"], MessageEmbed):
-            value["data"]["embed"] = type_validate_python(MessageEmbed, embed)
-        return cls(type="embed", data=value["data"])
+        if not isinstance(embed := instance.data["embed"], MessageEmbed):
+            instance.data["embed"] = type_validate_python(MessageEmbed, embed)
+        return instance
 
 
 class _ArkData(TypedDict):
@@ -348,19 +344,12 @@ class Ark(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if isinstance(value, cls):
-            return value
-        if isinstance(value, MessageSegment):
-            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
-        if not isinstance(value, dict):
-            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
-        if "data" not in value:
-            raise ValueError(f"Expected dict with 'data' for Ark, got {value}")
-        if "ark" not in value["data"]:
+        instance = super()._validate(value)
+        if "ark" not in instance.data:
             raise ValueError(f"Expected dict with 'ark' in 'data' for Ark, got {value}")
-        if not isinstance(ark := value["data"]["ark"], MessageArk):
-            value["data"]["ark"] = type_validate_python(MessageArk, ark)
-        return cls(type="ark", data=value["data"])
+        if not isinstance(ark := instance.data["ark"], MessageArk):
+            instance.data["ark"] = type_validate_python(MessageArk, ark)
+        return instance
 
 
 class _ReferenceData(TypedDict):
@@ -379,23 +368,16 @@ class Reference(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if isinstance(value, cls):
-            return value
-        if isinstance(value, MessageSegment):
-            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
-        if not isinstance(value, dict):
-            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
-        if "data" not in value:
-            raise ValueError(f"Expected dict with 'data' for Reference, got {value}")
-        if "reference" not in value["data"]:
+        instance = super()._validate(value)
+        if "reference" not in instance.data:
             raise ValueError(
                 f"Expected dict with 'reference' in 'data' for Reference, got {value}"
             )
-        if not isinstance(reference := value["data"]["reference"], MessageReference):
-            value["data"]["reference"] = type_validate_python(
+        if not isinstance(reference := instance.data["reference"], MessageReference):
+            instance.data["reference"] = type_validate_python(
                 MessageReference, reference
             )
-        return cls(type="reference", data=value["data"])
+        return instance
 
 
 class _MarkdownData(TypedDict):
@@ -414,21 +396,14 @@ class Markdown(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if isinstance(value, cls):
-            return value
-        if isinstance(value, MessageSegment):
-            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
-        if not isinstance(value, dict):
-            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
-        if "data" not in value:
-            raise ValueError(f"Expected dict with 'data' for Markdown, got {value}")
-        if "markdown" not in value["data"]:
+        instance = super()._validate(value)
+        if "markdown" not in instance.data:
             raise ValueError(
                 f"Expected dict with 'markdown' in 'data' for Markdown, got {value}"
             )
-        if not isinstance(markdown := value["data"]["markdown"], MessageMarkdown):
-            value["data"]["markdown"] = type_validate_python(MessageMarkdown, markdown)
-        return cls(type="markdown", data=value["data"])
+        if not isinstance(markdown := instance.data["markdown"], MessageMarkdown):
+            instance.data["markdown"] = type_validate_python(MessageMarkdown, markdown)
+        return instance
 
 
 class _KeyboardData(TypedDict):
@@ -447,21 +422,14 @@ class Keyboard(MessageSegment):
     @classmethod
     @override
     def _validate(cls, value) -> Self:
-        if isinstance(value, cls):
-            return value
-        if isinstance(value, MessageSegment):
-            raise ValueError(f"Type {type(value)} can not be converted to {cls}")
-        if not isinstance(value, dict):
-            raise ValueError(f"Expected dict for MessageSegment, got {type(value)}")
-        if "data" not in value:
-            raise ValueError(f"Expected dict with 'data' for Keyboard, got {value}")
-        if "keyboard" not in value["data"]:
+        instance = super()._validate(value)
+        if "keyboard" not in instance.data:
             raise ValueError(
                 f"Expected dict with 'keyboard' in 'data' for Keyboard, got {value}"
             )
-        if not isinstance(keyboard := value["data"]["keyboard"], MessageKeyboard):
-            value["data"]["keyboard"] = type_validate_python(MessageKeyboard, keyboard)
-        return cls(type="keyboard", data=value["data"])
+        if not isinstance(keyboard := instance.data["keyboard"], MessageKeyboard):
+            instance.data["keyboard"] = type_validate_python(MessageKeyboard, keyboard)
+        return instance
 
 
 SEGMENT_TYPE_MAP: Dict[str, Type[MessageSegment]] = {

